### DISCUSSION TITLE
C# Extension Method 관련 질문

### DISCUSSION THREAD

**Discussion by fkdl0048** [fkdl0048/Mentoring-Project#49] [Q&A] [answered]
멘토님과 static관련해서 이야기 나눴던 부분이 Extension Method를 알아보던 중 생각이 나서 질문 올립니다.

실제 프로젝트에서도 Extension Method를 자주 사용하나요?  

일종의 static형태인데 사용에 대해서 어떻게 생각하시는지 궁금합니다..!

**Comment by jongfeel** [comment, marked as answer]
static method의 경우는 상태를 저장하지 않고 순수하게 함수형으로만 사용한다면 크게 문제는 없긴 합니다.

그런데 extension의 경우 객체 타입에 대해 메서드를 확장하는 거라
문법은 static이 들어가지만 사실은 그 객체를 사용해서 새로운 메서드를 만드는 것에 가깝습니다.

일반적으로는 잘 안쓰는 것 같고
이미 만들어진 class에 수정을 가하기는 싫지만 간단히 기능을 더 추가하고 싶을 때 사용하면 유용하긴 합니다.
그게 장점이자 단점인데 디버깅을 해야 하는 어셈블리가 달라지면 디버거가 동작 안할 수도 있습니다.
그래서 상태를 저장하거나 하는 구현보다는 그 메서드 자체로 실행하고 해결되는 순수 함수로 만들 필요가 있습니다.
또 메서드 형태를 띠므로 함수를 구현한 것이라 객체지향적인 설계 레벨에는 별로 도움이 되지는 않습니다.

결국 확장 메서드는 객체의 메서드가 아니라 static method를 구현한 것이므로
적당히 쓰면 좋을 수는 있지만 너무 남발하게 된다면 그리 좋지는 않긴 합니다.

그리고 extension method로 할 수 있는 거면 그냥 클래스 메서드로 구현하는게 더 나을 수 있습니다.

**Reply by fkdl0048** [reply]
static method를 구현한건지 메서드가 추가된건지 내부 기능이 이해가 안되서 궁금했는데 감사합니다.  

이런 문제가 생기지 않도록 애초에 설계를 잘 해야겠네요..
